src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java

```diff
@@ -614,34 +580,133 @@ static void checkApiRateLimit(@NonNull TaskListener listener, GitHub github)
+            connections.put(connectionId, record);
+            reverseLookup.put(record.gitHub, record);
+            return record;
+        }
+
+
+        private void release() {
+            if (this.usageCount <= 1) {
+                this.usageCount = 0;
+                this.lastUsed = System.currentTimeMillis();
+            } else {
+                this.usageCount -= 1;
+            }
+        }
+
+        /**
+         * Remove all unused connections from the connection cache and delete their cache folders.
+         *
+         * @param threshold connections should be removed if their the last used time was before
+         *                  this system time in milliseconds.
+         * @param removeUntracked if true, also delete any cache folders that do not have
+         *                        corresponding connections. This operation can be expensive so it
+         *                        should be run as rarely as possible.
+         */
+        private static void removeAllUnused(long threshold, boolean removeUntracked) {
+            Map<String, File> validCaches = new HashMap<>();
+            synchronized (connections) {
+                for (Iterator<Map.Entry<ConnectionId, GitHubConnection>> iterator = connections.entrySet()
+                    .iterator();
                      iterator.hasNext(); ) {
-                    Map.Entry<GitHub, Long> entry = iterator.next();
-                    Long lastUse = entry.getValue();
-                    if (lastUse == null || lastUse < threshold) {
-                        iterator.remove();
-                        unused(entry.getKey());
+                    Map.Entry<ConnectionId, GitHubConnection> entry = iterator.next();
+                    try {
+                        GitHubConnection record = Objects.requireNonNull(entry.getValue());
+                        if (record.cache != null) {
+                            long lastUse = record.lastUsed;
+                            if (record.usageCount == 0 && lastUse < threshold) {
+                                iterator.remove();
+                                reverseLookup.remove(record.gitHub);
+                                if (record.cleanupCacheFolder) {
+                                    record.cache.delete();
+                                    record.cache.close();
+                                }
+                            } else if (removeUntracked) {
+                                validCaches.put(record.cache.directory().getAbsolutePath(),
+                                    record.cache.directory());
+                            }
+                        }
+                    } catch (IOException | NullPointerException e) {
+                        LOGGER.log(WARNING, "Exception for cache directory: " + entry.getKey(), e);
+                    }
+                }
+            }
+
+            if (removeUntracked) {
```

Comment:
Added this in the last change.  It is likely the existing users of github app credentials will have a large number of untracked cache folders.  This will attempt to clean them out.